simulariumio/custom_converter.py

```diff
@@ -92,3 +119,276 @@ def _read(self, input_data: CustomData) -> Dict[str, Any]:
             "data": input_data.plots,
         }
         return simularium_data
+
+    @staticmethod
+    def _get_spatial_bundle_data_subpoints(
+        agent_data: AgentData, used_unique_IDs: List[int] = []
+    ) -> List[Dict[str, Any]]:
+        """
+        Return the spatialData's bundleData for a simulation
+        of agents with subpoints, packing buffer with jagged data is slower
+        """
+        bundleData: List[Dict[str, Any]] = []
+        uids = {}
+        for t in range(len(agent_data.times)):
+            # timestep
+            frame_data = {}
+            frame_data["frameNumber"] = t
+            frame_data["time"] = float(agent_data.times[t])
+            n_agents = int(agent_data.n_agents[t])
+            i = 0
+            buffer_size = (len(V1_SPATIAL_BUFFER_STRUCT) - 1) * n_agents
+            for n in range(n_agents):
+                s = int(agent_data.n_subpoints[t][n])
+                if s > 0:
+                    buffer_size += 3 * s
+                    if agent_data.draw_fiber_points:
+                        buffer_size += (len(V1_SPATIAL_BUFFER_STRUCT) - 1) * max(
+                            math.ceil(s / 2.0), 1
+                        )
+            local_buf = np.zeros(buffer_size)
+            for n in range(n_agents):
+                # add agent
+                local_buf[
+                    i + V1_SPATIAL_BUFFER_STRUCT.index("VIZ_TYPE")
+                ] = agent_data.viz_types[t, n]
+                local_buf[
+                    i + V1_SPATIAL_BUFFER_STRUCT.index("UID")
+                ] = agent_data.unique_ids[t, n]
+                local_buf[
+                    i + V1_SPATIAL_BUFFER_STRUCT.index("TID")
+                ] = agent_data.type_ids[t, n]
+                local_buf[
+                    i
+                    + V1_SPATIAL_BUFFER_STRUCT.index("POSX") : i
+                    + V1_SPATIAL_BUFFER_STRUCT.index("POSX")
+                    + 3
+                ] = agent_data.positions[t, n]
+                local_buf[i + V1_SPATIAL_BUFFER_STRUCT.index("R")] = agent_data.radii[
+                    t, n
+                ]
+                n_subpoints = int(agent_data.n_subpoints[t][n])
+                if n_subpoints > 0:
+                    # add subpoints to fiber agent
+                    subpoints = [3 * n_subpoints]
+                    for p in range(n_subpoints):
+                        for d in range(3):
+                            subpoints.append(agent_data.subpoints[t][n][p][d])
+                    local_buf[
+                        i
+                        + V1_SPATIAL_BUFFER_STRUCT.index("NSP") : i
```

Comment:
if V1_SPATIAL_BUFFER_STRUCT never changes, all these `.index` calls could be symbolic constants.  Or at least the hardcoded strings could be.

Reply:
what would V1_SPATIAL_BUFFER_STRUCT look like then?